src/ngraph/runtime/cpu/pass/cpu_reshape_sinking.cpp

```diff
@@ -0,0 +1,260 @@
+            NGRAPH_DEBUG << "Propagating " << describe_reshape(arg_reshape) << " for "
+                         << n->get_name();
+            reorders[n] = reorders[n->get_argument(0)];
+        }
+        else if (auto binary = std::dynamic_pointer_cast<op::util::BinaryElementwiseArithmetic>(n))
+        {
+            auto left = n->get_argument(0);
+            auto right = n->get_argument(1);
+
+            if (reorders.at(left)->get_input_order() == reorders.at(right)->get_input_order())
+            {
+                NGRAPH_DEBUG << "Propagating " << describe_reshape(reorders.at(left)) << " for "
+                             << n->get_name();
+                reorders[n] = reorders.at(left);
+            }
+            else if (reorders.at(left)->get_input_order() ==
+                     ngraph::get_default_order(left->get_shape()))
+            {
+                convert_binary_to_default_order(
+                    binary, binary->get_inputs().at(0), right, reorders);
+            }
+            else if (reorders.at(right)->get_input_order() ==
+                     ngraph::get_default_order(right->get_shape()))
+            {
+                convert_binary_to_default_order(binary, binary->get_inputs().at(1), left, reorders);
+            }
+            else
+            {
+                NGRAPH_DEBUG << "Materializing both reshapes for " << binary->get_name();
+                NGRAPH_DEBUG << "Left = " << describe_reshape(reorders.at(left));
+                NGRAPH_DEBUG << "Right = " << describe_reshape(reorders.at(right));
+                delete_reshape(reorders.at(left));
+                delete_reshape(reorders.at(right));
+                insert_reshape(binary, reorders.at(left), 0);
+                insert_reshape(binary, reorders.at(right), 1);
+            }
+        }
+        else if (auto goe = std::dynamic_pointer_cast<op::GetOutputElement>(n))
+        {
+            reorders[goe] = create_default_reshape(goe);
+        }
+        else
+        {
+            //skip multiple output nodes and deal with GOEs exclusively
+            if (n->get_outputs().size() > 1)
+            {
+                continue;
+            }
+            //TODO: multiple outputs
+            for (size_t i = 0; i < n->get_arguments().size(); i++)
+            {
+                //materialize all pending reshapes, flush pending reshapes
+                auto arg = n->get_argument(i);
+                if (reorders.count(arg) != 0)
+                {
+                    NGRAPH_DEBUG << "Materializing " << describe_reshape(reorders.at(arg))
+                                 << " for " << arg->get_name();
+                    delete_reshape(reorders.at(arg));
+                    insert_reshape(n, reorders.at(arg), i);
+                    //no swimming up
+                }
+            }
+            reorders[n] = create_default_reshape(n);
+        }
+    }
+
+    //fix wrong shape info wholesale
```

Comment:
why we are fixing shapes? it will allow us to serialize, copy a graph after `ReshapeSinking`. Also, some optimizations (`ReshapeElimination` and most likely `CPULayout`) rely on correct shape info.